### PR TITLE
NEW Add hook for virtual stock

### PR DESCRIPTION
**Pull request by defrance**
Allow to add/substract some qty for GPAO OF running

### CODE REVIEW

File: htdocs/product/class/product.class.php [hunk 1]
```diff
@@ -3975,6 +3975,16 @@ function load_virtual_stock()
         if (! empty($conf->global->STOCK_CALCULATE_ON_SUPPLIER_BILL)) {
             $this->stock_theorique+=$stock_commande_fournisseur-$stock_reception_fournisseur;
         }
+
+	// Call triggers
+	include_once DOL_DOCUMENT_ROOT . '/core/class/interfaces.class.php';
+	$interface=new Interfaces($this->db);
+	$result=$interface->run_triggers('LOAD_VIRTUAL_STOCK', $this, $user, $langs, $conf);
```
Comment (eldy) [inline htdocs/product/class/product.class.php]: Loading data is not a business event on workflow (not a Create/Update/Delete event), so usage of trigger is not allowed here.
May be you can address your need with a hook.

Comment (defrance) [reply htdocs/product/class/product.class.php]: Hook is for the display (Vue) , Trigger for action (Controler), but nothing for the data (Model)
My goal is to increase or decrease the virtual stock based on some other data rule

Comment (defrance) [reply htdocs/product/class/product.class.php]: i have search some other way, completesubstitutionarray is not a good one

Comment (hregis) [reply htdocs/product/class/product.class.php]: @defrance 
you must be able to overload the fetch in the "doActions" hook of the page

Comment (defrance) [reply htdocs/product/class/product.class.php]: yes but in this case it's a global hook and the list of virtual stock is not good.
For me the hook is for change the Vue but not the model and it's more "clean" to work with trigger to do that

Comment (eldy) [reply htdocs/product/class/product.class.php]: The code to change controller is the doAction hook, trigger is an expensive process that must be used for CUD events (create, update, delete). They are never called in just browsing data.

See https://wiki.dolibarr.org/index.php/Hooks_system
and https://wiki.dolibarr.org/index.php/Triggers

**Comment by defrance** [comment]
Trigger was remplaced by hook

**Pull request merged**